subxt/src/events/events_client.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use crate::{
+    client::OnlineClientT,
+    error::Error,
+    events::{
+        EventSub,
+        EventSubscription,
+        Events,
+        FinalizedEventSub,
+    },
+    Config,
+};
+use derivative::Derivative;
+use futures::{
+    future::Either,
+    stream,
+    Stream,
+    StreamExt,
+};
+use sp_core::{
+    storage::StorageKey,
+    twox_128,
+};
+use sp_runtime::traits::Header;
+use std::future::Future;
+
+/// A client for working with events.
+#[derive(Derivative)]
+#[derivative(Clone(bound = "Client: Clone"))]
+pub struct EventsClient<T, Client> {
+    client: Client,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T, Client> EventsClient<T, Client> {
+    /// Create a new [`EventsClient`].
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T, Client> EventsClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Obtain events at some block hash.
+    pub fn at(
+        &self,
+        block_hash: Option<T::Hash>,
+    ) -> impl Future<Output = Result<Events<T>, Error>> + Send + 'static {
```

Comment:
Just because anm `async fn` would have an implicit lifetime tied to self; here we clone the client out of self so that the returned future can have a `'static` lifetime. I think the motivating example here was so you could do stuff like `let fut = api.events().at(...)` and then `futures::join!(fut, fut2..)`; with a lifetime tied to self I think it would complain that `api.events()` was temporary and didn't live long enough if self was needlessly borrowed from it

Reply:
ah, I see